### PR TITLE
Add language selector to change the app language

### LINKED ISSUE #1130: Do not use IP to select app's language
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.
- [x] I checked if the issue/feature exists in the latest version.

Currently it seems NewPipe uses the IP to determine the language of the app. This is pretty sub-optimal for any user that uses an out-of-country VPN, or for users who wish to use another language (I set all my applications to English, for example). It's also harder to debug translations thi way. Could it be made so that it either has a language selector, or selects the language depending on the system locale?

### LINKED ISSUE #1215: 0.13.0 beta - Polish and English are both displayed at the same time
Hi, 

I checked the issue tracker but I couldn't find if my issue exists or my github-fu skills are lacking.

In the 0.13.0 beta for NewPipe there are some Polish words mixed in with English when I open up NewPipe.

https://i.imgur.com/DuhQwM1.jpg

If you look at the attached screenshot, you will see the words "2 dni temu" (2 days ago) mixed in with everything else.

My phone's language is set to English (United States) and my input languages are set English (United States) and Polish. My region in my Android settings is set to Poland. The content country in my install of NewPipe is set to United Kingdom. Figured I should mention that in case any of what I just mentioned has a bearing on this issue.

Let me know if you need any further info.

### LINKED ISSUE #2162: [feature request] UI Language selection menu or "Force English" in settings
I can not find a way to switch from one language to another, is there any?

### LINKED ISSUE #2794: Change languages settings' behaviour
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.
- [x] I checked if the issue/feature exists in the latest version.

Hi,
I just finished the translation of Esperanto (I didn't translate everything alone, thank you other translators) on Weblate and noticed that there is a problem with it.

Currently, the languages settings' behaviour is the following:
− "Default content country" especially changes Trendings page (I didn't see what else).
− I have no clue what default content language changes.
− There is no way to change the app's language from the app, you must change your phone's app.

The last point is the problem, as Esperanto is a constructed international auxiliary language, almost everyone who speak it also speak another language (the ~2000 native speakers are bilingual (or even more), and the other speakers — about ten millions in the world — also speak their native language);
and isn't translated on every software, including phones' OS.
Therefore you won't be able to change NewPipe into Esperanto even though the community made an effort to translate to Esperanto.

What I think we should have in NewPipe: 
− By default, NewPipe's language is your phone's language (as it's now).
− "Default content language" changes NewPipe's language.
Or a third language option but that would confuse people, then you should clarify what they actually do in the settings, so that people aren't lost.

Also, reCAPTCHA string should not be let translatable (as NewPipe, MD5 or SHA-1 is), because it's causing "Failed check: Unchanged translation"

### LINKED ISSUE #2887: translation is work when change app into different language.
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.
- [x] I checked if the issue/feature exists in the latest version.
- [x] I did use the [incredible bugreport to markdown converter](https://teamnewpipe.github.io/CrashReportToMarkdown/) to paste bug reports.
translation is work when change app into different language. Like Hindi, chines etc.

### PR DESCRIPTION
**Pull request by B0pol**
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.

Fix #2794, fix #2162, fix #2887, fix #1215, fix #1130, fix [r/NewPipe : Any way to change the app language?](https://www.reddit.com/r/NewPipe/comments/b8sqpf/any_way_to_change_the_app_language/), [r/NewPipe : how to change app language ?](https://www.reddit.com/r/NewPipe/comments/elamrg/how_to_change_app_language/)

Explaination :
Before, NewPipe's language was phone's language. Now, it's the language selected in Settings -> Content -> App language, and by default, it stays system default language.
Thus, you can force an app language (eg. English for developing / debugging : take a screenshot in English) or use non-available system language (Esperanto isn't translated for Android) or simply just set the language you want **without changing phone / system language**.

APK: [2020-01-27_language-selector#2921_APK.zip](https://github.com/TeamNewPipe/NewPipe/files/4118734/2020-01-27_language-selector.2921_APK.zip)

### CODE REVIEW

**Comment by B0pol** [comment]
I found out what does content language and also I found it's not setting the language you want sometimes :
If the language is like language-COUNTRY, for example en-GB (UK english), it doesn't work and use phone's language. Same if the language in the list is not translated : actually, the list is used for search and other stuff. 
Then think we should use a third selector, see this : 
Screenshot_1578325520
where we would set the language already translated in NewPipe, + at the top a button that would be : system language.

I'll work on it this week

**Comment by B0pol** [comment]
I finished the work and tested all languages. I found an exception (#2940), but it's not related to my PR, it's related to the formatter and %1$s not in the good order and arabic alphabet, then you can merge this PR.

**Comment by B0pol** [comment]
I changed the way time is displayed, it now use NewPipe's language (the new selector).
This fixes #1215 . In the limit of the languages of the library Pretty time we use : https://www.ocpsoft.org/prettytime/#section-12
I'll try to use custom TimeFormat with Esperanto (the library supports to add languages), and if it works, it will be translatable with weblate for all the other languages.

**Comment by B0pol** [comment]
The PR is ready for a review, please consider doing it @Stypox @TobiGr @mauriciocolli
BTW, here is some screenshots :
System language was Spanish
[US trendings displayed in English](https://user-images.githubusercontent.com/58657617/72161936-ebb92580-33b8-11ea-927c-9f88d26f4f5b.png) | [Settings in English](https://user-images.githubusercontent.com/58657617/72162178-5bc7ab80-33b9-11ea-98e9-30143aa4afe8.png)

[US trendings displayed in German](https://user-images.githubusercontent.com/58657617/72161986-07243080-33b9-11ea-9640-0e13a31e6eaf.png) | [Settings in German](https://user-images.githubusercontent.com/58657617/72162018-15724c80-33b9-11ea-828e-e1145627a244.png)

[US trendings displayed in System language, Spanish](https://user-images.githubusercontent.com/58657617/72162273-92052b00-33b9-11ea-966d-e364c44331f4.png) | [Settings in System language, Spanish](https://user-images.githubusercontent.com/58657617/72162134-4b173580-33b9-11ea-9132-938985540e18.png)

**Comment by B0pol** [comment]
I just fixed Indonesian that wasn't possible to select by renaming values-id to values-in.
Briefly, android use outdated version of an iso norm, then it doesn't recognize id, but in.
See https://stackoverflow.com/questions/13291578/how-to-localize-an-android-app-in-indonesian-language.
I also added Occitan to the selector.
So the selector is working for every language now.
Pirate will show English as [values-pr/strings.xml](https://github.com/TeamNewPipe/NewPipe/blob/dev/app/src/main/res/values-pr/strings.xml) is empty, and for some languages you will see a lot of english, as it's not translated much (I'm thinking to Interlingua, Occitan…)

**Comment by B0pol** [comment]
APK: [app-debug-language-selector_2020-01-18.zip](https://github.com/TeamNewPipe/NewPipe/files/4080229/app-debug-language-selector_2020-01-18.zip)

**Comment by B0pol** [comment]
@abusarimhindi APK: [apk-debug](https://github.com/TeamNewPipe/NewPipe/files/4108063/app-debug.zip)
In settings, you can change for Urdu in settings / content settings / NewPipe's language.

For anyone else wondering, this is a TEST version: there might be some problems, because I was based on dev branch in extractor, and I switched to master branch of extractor (the only one that includes the fix for decryption bug)

I manually updated extractor version so that YouTube works again, and remove the urdu string, took the one from weblate.

**Comment by abusarimhindi** [comment]
> APK: [app-debug-language-selector_2020-01-19.zip](https://github.com/TeamNewPipe/NewPipe/files/4082971/app-debug-language-selector_2020-01-19.zip)
Thanks!
But، I think this is not updated version (0.18.2), and Urdu updated strings are not putted in this file.

Annotation 2020-01-24 175654

**Comment by B0pol** [comment]
This is 0.18.1 because I'm based on debug build, which is still on 0.18.1.
But as I said, I manually updated extractor version, to prevent the decryption crash, and I manually updated urdu strings from weblate, just for you, and I didn't commit these changes because this should be fixed by updating dev branch, not by fixing manually

**Comment by abusarimhindi** [comment]
> This is 0.18.1 because I'm based on debug build, which is still on 0.18.1.
> But as I said, I manually updated extractor version, to prevent the decryption crash, and I manually updated urdu strings from weblate, just for you, and I didn't commit these changes because this should be fixed by updating dev branch, not by fixing manually

Sorry, It is Ok. Firstly I downloaded the old file, now updated. Thanks again.

File: app/src/main/java/org/schabi/newpipe/util/Localization.java [hunk 1]
```diff
@@ -199,21 +204,41 @@ public static String getDurationString(long duration) {
...
+        res.updateConfiguration(conf, dm);
+    }
+
+    public static Locale getAppLocale(Context context) {
+        SharedPreferences prefs = androidx.preference.PreferenceManager.getDefaultSharedPreferences(context);
+        String lang = prefs.getString("newpipes_language_key", "en");
+        Locale loc;
+        if (lang.equals("system")) {
+            loc = Locale.getDefault();
+        } else if (lang.matches(".*-.*")) {
```
[the remark this answers is not in the document]
Comment (B0pol) [reply app/src/main/java/org/schabi/newpipe/util/Localization.java]: not en-gb and en-us (there is only one form of english in NewPipe and weblate), but pt (portugal) from pt-br (brazil), same for nl and nl-be, chinese… I'll do this in the next commit

**Comment by B0pol** [comment]
> ```java
> public static void assureCorrectAppLanguage() {
>     changeAppLanguage(getAppLocale(getApplicationContext()), getResources());
> }
> ```

I can't do that, I can't have context and resources from a static context.
The most optimized thing I can have is
```java
public static void assureCorrectAppLanguage(Context c) {
    changeAppLanguage(getAppLocale(c), c.getResources())
}
```
and then, in mainactivity and other activites needing it:
``assureCorrectAppLanguage(this);`` or ``assureCorrectAppLanguage(getContext());`` sometimes.
As assureCorrectAppLanguage(getApplicationContext()); doesn't work, ill need time for testing.

> Is it necessary to call changeAppLanguage() in all onCreate and onResume methods or can we just do it in App.java in?

Yesn't sir! If you put in onCreate and onResume of mainActivity, it will work for about 95%. I'll say why I put for the other files:
AboutActivity: everything is fine because strings are loaded on runtime, except title which is set in manifest.xml, so I changed it on runtime

Player: kinda tough, it is in the language you want, except if you press a button, change orientation (landscape / portrait).

The other file honestly I don't remember, but at first I put in MainActivity, and while testing phase I added where it would show the wrong language.
It's kinda like the ``ThemeHelper.setTheme(this, ServiceHelper.getSelectedServiceId(this));``: you don't need _everywhere_, but in activities who are making problems without it.

I've tested for a long time since I opened the PR, then I'm almost sure there won't be those issues with the current build, but if someone found one later, you just need to add ``assureCorrectAppLanguage(getApplicationContext());``, and as I said, it works most of the time, as I said by putting only in MainActivity it works almost everywhere.

**Comment by TobiGr** [comment]
> > ```java
> > public static void assureCorrectAppLanguage() {
> >     changeAppLanguage(getAppLocale(getApplicationContext()), getResources());
> > }
> > ```
> 
> I can't do that, I can't have context and resources from a static context.
> The most optimized thing I can have is
> 
> ```java
> public static void assureCorrectAppLanguage(Context c) {
>     changeAppLanguage(getAppLocale(c), c.getResources());
> }
> ```

My bad. This happens when you write reviews during lectures :D

**Comment by TobiGr** [comment]
Code looks go so far. I'll test it as soon as possible

**Comment by B0pol** [comment]
Just one thing you should know
I made changes to values-ur/strings.xml, because of formatter crash (#2940)
They were fixed by weblate, so either you update weblate soon and I solve conficts before you merge the pr, or for every conflicts existing with Urdu strings, you should pick the Weblate version

**Comment by TobiGr** [comment]
> I made changes to values-ur/strings.xml, because of formatter crash (#2940)
They were fixed by weblate in the meantime, so either you update dev with weblate soon and I solve conficts before you merge the pr, or for every future conflicts existing with Urdu strings, you should pick the Weblate version.

I merged weblate into dev. Could you take a look at the conflicts? I am short on time until I finished my exams. Thanks :)

**Comment by B0pol** [comment]
> I merged weblate into dev. Could you take a look at the conflicts? I am short on time until I finished my exams. Thanks :)

Done. Actually, the only conflict shown on Github was the indonesian file which have been renamed, but while merging with git I had no problem.

I think the PR is finally fully ready.

**Comment by TobiGr** [comment]
Thanks. One small thing. Can you please integrate this setting into the mechanism which displays a toast that language changes require an app restart?

https://github.com/TeamNewPipe/NewPipe/blob/c5e6bb58bce295a18e0d5a478a7c77eb0ee522b9/app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java#L135-L140

**Comment by B0pol** [comment]
> Thanks. One small thing. Can you please integrate this setting into the mechanism which displays a toast that language changes require an app restart?
> 
> https://github.com/TeamNewPipe/NewPipe/blob/c5e6bb58bce295a18e0d5a478a7c77eb0ee522b9/app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java#L135-L140

It's already done. Look at https://github.com/TeamNewPipe/NewPipe/blob/c5e6bb58bce295a18e0d5a478a7c77eb0ee522b9/app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java#L135-L136 and https://github.com/TeamNewPipe/NewPipe/blob/c5e6bb58bce295a18e0d5a478a7c77eb0ee522b9/app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java#L71

**Comment by B0pol** [comment]
> Thanks. One small thing. Can you please integrate this setting into the mechanism which displays a toast that language changes require an app restart?
> 
> https://github.com/TeamNewPipe/NewPipe/blob/c5e6bb58bce295a18e0d5a478a7c77eb0ee522b9/app/src/main/java/org/schabi/newpipe/settings/ContentSettingsFragment.java#L135-L140

done

**Pull request merged**